File: c++/73_Set_Matrix_Zeroes.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
    	int m = matrix.size();
    	int n = matrix[0].size();
    	int tagRow[m] = {0};
    	int tagCOl[n] = {0};
    	for (int i = 0; i < m; i++) {
    		for (int j = 0; j < n; j++) {
    			if (matrix[i][j] != 0)
    				continue;
    			tagRow[i] = 1;
    			tagCOl[j] = 1;
    		}
    	}
    	for (int i = 0; i < m; i++)
    		for (int j = 0; j < n; j++)
    			if (tagRow[i] || tagCOl[j])
    				matrix[i][j] = 0;    
    }
};

// there is actually a better way which uses o(1) space
// the trick is to mark infos in the first row and col


int main() {
	Solution s;
}

File: python/73_Set_Matrix_Zeroes.py
#!/usr/bin/python


class Solution(object):
    def setZeroes(self, matrix):
        m = len(matrix)
        n = len(matrix[0])
        tagRow = [0] * m
        tagCol = [0] * n
        for i in range(m):
        	for j in range(n):
        		if matrix[i][j] == 0:
        			tagRow[i], tagCol[j] = 1, 1
       	for i in range(m):
        	for j in range(n):
        		if tagRow[i] or tagCol[j]:
        			matrix[i][j] = 0
        

def main():
    s = Solution()

if __name__ == "__main__":
    main()
